### ISSUE TITLE
BugData unique index

### ISSUE THREAD

**Issue by philipbaileynar** [namc-utah/namc-bug-db#42]
We have discussed in #30 how we intend to have a unique index on `PilotDB.BugData` across the following fields:

- SampleID
- Code
- LifeStageID
- BugSize

@David-Fowler mentioned the cases where there are no bug sizes and I have come across a sample with 94 bug identifications all of which have bug size = 0. These break the unique index described above.

```sql
SELECT * FROM BugData WHERE SampleID = 152859
```

Is BugSize zero valid? Or do these represent the absence of a bug size, in which case they should have BugSize set to NULL?

## Options

1. Drop the unique index altogether.
2. Limit the index to just those records that have bug sizes (referred to as a partial index).
2. Keep the unique index and combine duplicate records by summing all bug counts and big rare counts  with the same life stage and species within a sample.

## Supporting Information

Here are the first 10 records of the 94 related to this sample:

| SampleID | Code | RowIndex | LifeStage | BugSize | SplitCount | BigRareCount | Notes | Valid |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 152859 | 184 | NULL | L | 0 | 100 | 2 | NULL | NULL |
| 152859 | 182 | NULL | L | 0 | 53 | 0 | NULL | NULL |
| 152859 | 180 | NULL | P | 0 | 5 | 0 | NULL | NULL |
| 152859 | 216 | NULL | L | 0 | 5 | 0 | NULL | NULL |
| 152859 | 223 | NULL | L | 0 | 12 | 0 | NULL | NULL |
| 152859 | 908 | NULL | L | 0 | 9 | 0 | NULL | NULL |
| 152859 | 201 | NULL | L | 0 | 8 | 0 | NULL | NULL |
| 152859 | 706 | NULL | L | 0 | 2 | 0 | NULL | NULL |
| 152859 | 235 | NULL | L | 0 | 29 | 3 | NULL | NULL |
| 152859 | 241 | NULL | L | 0 | 1 | 0 | NULL | NULL |

And here are the 36 BugData records where there are more than one occurence of SampleID, Code, LifeStageID, BugSize:

| SampleID | Code | LifeStage | BugSize |  |
| :--- | :--- | :--- | :--- | :--- |
| 152859 | 124 | L | 0 | 2 |
| 152859 | 201 | L | 0 | 2 |
| 152859 | 223 | L | 0 | 2 |
| 152859 | 235 | L | 0 | 2 |
| 152859 | 262 | L | 0 | 2 |
| 152859 | 280 | L | 0 | 2 |
| 152859 | 491 | L | 0 | 2 |
| 152859 | 499 | L | 0 | 2 |
| 152859 | 121 | L | 0 | 2 |
| 152859 | 135 | L | 0 | 2 |
| 152859 | 267 | L | 0 | 2 |
| 152859 | 462 | L | 0 | 2 |
| 152859 | 484 | L | 0 | 2 |
| 152859 | 646 | A | 0 | 2 |
| 152859 | 706 | L | 0 | 2 |
| 152859 | 19 | A | 0 | 2 |
| 152859 | 58 | A | 0 | 2 |
| 152859 | 146 | L | 0 | 2 |
| 152859 | 180 | P | 0 | 2 |
| 152859 | 182 | L | 0 | 2 |
| 152859 | 184 | L | 0 | 2 |
| 152859 | 247 | L | 0 | 2 |
| 152859 | 287 | L | 0 | 2 |
| 152859 | 807 | L | 0 | 2 |
| 152859 | 908 | L | 0 | 2 |
| 152859 | 216 | L | 0 | 2 |
| 152859 | 237 | L | 0 | 2 |
| 152859 | 279 | L | 0 | 2 |
| 152859 | 412 | L | 0 | 2 |
| 152859 | 519 | L | 0 | 2 |
| 152859 | 550 | L | 0 | 2 |
| 152859 | 778 | L | 0 | 2 |
| 152859 | 808 | L | 0 | 2 |
| 152859 | 809 | L | 0 | 2 |
| 152859 | 970 | L | 0 | 2 |
| 152859 | 1274 | L | 0 | 2 |

```sql
SELECT SampleID, Code, LifeStage, BugSize, Count(*) FROM BugData WHERE SampleID = 152859 GROUP BY SampleId, Code, LifeStage, BugSize HAVING Count(*) > 1
```

**Comment by David-Fowler** [comment]
I am not at my computer but when I get back I will check the full contents of that box against measurements. More than likely the 0 will be changed to null and our indexing will stay as we discussed.

**Comment by Feldspar5** [comment]
Sorry for the delayed response; I am enjoying the time away too much!

The vast majority of bugs are not measured for size in our db.  Size should not be used as a unique index value.  Bugs with size 0 should be set to NULL.

Trip

**Comment by philipbaileynar** [comment]
This issue is resolved by #195
